src/create-template-orders.js

```diff
@@ -20,20 +33,263 @@ async function _fetchLastStartTimestamp() {
+    },
+  ]
+  await apiRoot
+    .orders()
+    .withId({ ID: checkoutOrder.id })
+    .post({
+      body: {
+        actions: updateActions,
+        version: checkoutOrder.version,
+      },
+    })
+    .execute()
+}
+
+async function _processCheckoutOrder(checkoutOrder) {
+  try {
+    const templateOrderDrafts = _generateTemplateOrderDrafts(checkoutOrder)
+
+    await pMap(
+      templateOrderDrafts,
+      async (templateOrderDraft) => {
+        await _createTemplateOrderAndPayments(checkoutOrder, templateOrderDraft)
+      },
+      { concurrency: 3 }
+    )
+
+    await _setCheckoutOrderProcessed(checkoutOrder)
+  } catch (err) {
+    logger.error(
+      `Failed to create template order from the checkout order with number ${checkoutOrder.orderNumber}. ` +
+        'Skipping this checkout order' +
+        ` Error: ${JSON.stringify(serializeError(err))}`
+    )
+  }
+}
+
+function isDuplicateOrderError(e) {
+  return (
+    e.code === 400 &&
+    e.body?.errors?.length === 1 &&
+    e.body?.errors?.some(
+      (err) => err.code === 'DuplicateField' && err.field === 'orderNumber'
+    )
+  )
+}
+
+function _generateTemplateOrderDrafts(checkoutOrder) {
+  const orderDrafts = []
+  const subscriptionLineItems = checkoutOrder.lineItems.filter(
+    (lineItem) => lineItem.custom?.fields?.isSubscription
+  )
+  for (const lineItem of subscriptionLineItems)
+    if (lineItem.quantity > 1)
+      for (let i = 0; i < lineItem.quantity; i++) {
+        const templateOrderDraft = _generateTemplateOrderImportDraft(
+          checkoutOrder,
+          lineItem,
+          i + 1
+        )
+        orderDrafts.push(templateOrderDraft)
+      }
+    else {
+      const templateOrderDraft = _generateTemplateOrderImportDraft(
+        checkoutOrder,
+        lineItem
+      )
+      orderDrafts.push(templateOrderDraft)
+    }
+
+  return orderDrafts
+}
+
+function _generateTemplateOrderImportDraft(
+  checkoutOrder,
+  lineItem,
+  quantityIncrement
+) {
+  const schedule = lineItem.custom.fields.schedule
+  const cronExpression = parser.parseExpression(schedule)
+  let nextDeliveryCronDate = cronExpression.next()
+  const nextDeliveryDate = nextDeliveryCronDate.toDate()
+  let nextDeliveryDateISOString = nextDeliveryDate.toISOString()
+  const cutoffDays = lineItem.custom.fields.cutoffDays
+  if (cutoffDays) {
+    nextDeliveryDate.setDate(nextDeliveryDate.getDate() - cutoffDays)
+    if (nextDeliveryDate.toISOString() < checkoutOrder.createdAt) {
+      nextDeliveryDateISOString = nextDeliveryDate.toISOString()
+      nextDeliveryCronDate = cronExpression.next()
+    }
+  }
+  const reminderDays = lineItem.custom.fields.reminderDays
+  let nextReminderDateISOString
+  if (reminderDays) {
+    nextDeliveryCronDate.setDate(nextDeliveryCronDate.getDate() - reminderDays)
+    nextReminderDateISOString = nextDeliveryCronDate.toISOString()
+  }
+  const templateOrder = {
+    orderNumber:
+      lineItem.custom.fields.subscriptionKey +
```

Comment:
I guess this function require validations before preparing all draft, instead of prepare and then failure on the platform, we should not send unnecessary requests to platform, for instance here: if the subscriptionKey is empty, then all operations before is just unnecessary, with validation before we could provide more useful logs etc.

Reply:
You are right, but this is too much for the first iteration. All these could be done later, so let's discuss this once we finish our first release and have more time.